test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs

```diff
@@ -125,29 +86,104 @@ public void RecordAttachmentsShouldAddToAttachmentSet()
 
             Assert.IsNotNull(attachments);
             CollectionAssert.AreEqual(attachmentSet, attachments);
         }
+
+        #region TestCaseResult caching tests.
+        [TestMethod]
+        public void RecordStartShouldInvokeSendTestCaseStart()
+        {
+            this.testRecorderWithTestEventsHandler.RecordStart(this.testCase);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestCaseStart(this.testCase), Times.Once);
+        }
+
+        [TestMethod]
+        public void SendTestCaseEndShouldInovkeTestCaseEndEventIfTestCaseStartWasCalledBefore()
+        {
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase,TestOutcome.Passed);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestCaseEnd(this.testCase, TestOutcome.Passed), Times.Once);
+        }
+
+        [TestMethod]
+        public void SendTestCaseEndShouldNotInvokeTestCaseEndEventInCaseOfAMissingTestCaseStartInDataDrivenScenario()
+        {
+            this.testRecorderWithTestEventsHandler.RecordStart(this.testCase);
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase,TestOutcome.Passed);
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase,TestOutcome.Failed);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestCaseEnd(this.testCase, TestOutcome.Passed), Times.Once);
+        }
+
+        [TestMethod]
+        public void RecordEndShouldInvokeSendTestCaseEndMultipleTimesInDataDrivenScenario()
+        {
+            this.testRecorderWithTestEventsHandler.RecordStart(this.testCase);
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase,TestOutcome.Passed);
+            this.testRecorderWithTestEventsHandler.RecordStart(this.testCase);
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase,TestOutcome.Failed);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestCaseEnd(this.testCase, TestOutcome.Passed), Times.Once);
+        }
+
+        [TestMethod]
+        public void RecordResultShouldPublishTestCaseResultEventIfTestCaseStartAndTestCaseEndEventsAreNotPublished()
+        {
+            this.testRecorderWithTestEventsHandler.RecordResult(this.testResult);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestResult(this.testResult), Times.Never);
+        }
+
+        [TestMethod]
+        public void RecordResultShouldPublishTestCaseResultEventIfTestCaseStartAndTestCaseEndEventsArePublished()
+        {
+            this.testRecorderWithTestEventsHandler.RecordStart(this.testCase);
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase, TestOutcome.Passed);
+            this.testRecorderWithTestEventsHandler.RecordResult(this.testResult);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestResult(testResult), Times.Once);
+        }
+
+        [TestMethod]
+        public void RecordResultShouldFlushIfTestCaseEndWasCalledBefore()
```

Comment:
We need tests to validate the order of calls: TestCaseStart, TestCaseResult (1 or more), TestCaseEnd